internal/exec/stages/files/units.go

```diff
@@ -64,10 +102,72 @@ func (s *stage) createUnits(config types.Config) error {
 			s.relabel(relabelpath)
 		}
 	}
-	// and relabel the preset file itself if we enabled/disabled something
-	if enabledOneUnit {
-		s.relabel(util.PresetPath)
+	// if we have presets then create the systemd preset file.
+	if len(presets) != 0 {
+		if err := s.createSystemdPresetFile(presets); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// parseInstanceUnit is a helper function to extract the name
+// and coresponding instances for a given instantiated unit.
+func parseInstanceUnit(unit types.Unit) (string, string, error) {
+	at := strings.Index(unit.Name, "@")
+	dot := strings.LastIndex(unit.Name, ".")
+	if dot == -1 {
+		return "", "", errors.ErrNoDotSuffix
+	}
+	instance := unit.Name[at+1 : dot]
+	serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
+	return serviceInstance, instance, nil
+}
+
+// createSystemdPresetFile creates the presetfile for enabled/disabled
+// instantiated/systemd units.
+func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
+	var enableUnitString, disableUnitString string
+	hasInstanceUnit := false
+	for _, value := range presets {
+		if value.instantiateable {
+			hasInstanceUnit = true
+		}
+		if value.enabled {
+			if value.instantiateable {
+				enableUnitString = value.unit + " " + strings.Join(value.instances, " ")
+			} else {
+				enableUnitString = value.unit
+			}
+			if err := s.Logger.LogOp(
+				func() error { return s.EnableUnit(enableUnitString) },
+				"enabling unit %q", enableUnitString,
+			); err != nil {
+				return err
+			}
+		} else {
+			if value.instantiateable {
+				disableUnitString = value.unit + " " + strings.Join(value.instances, " ")
+			} else {
+				disableUnitString = value.unit
+			}
+			if err := s.Logger.LogOp(
+				func() error { return s.DisableUnit(disableUnitString) },
+				"disabling unit %q", disableUnitString,
+			); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Print the warning if there's an instantiated unit present under
+	// the systemd units and the version of systemd in a given system
+	// is older than 240.
+	if hasInstanceUnit {
+		s.warnOnOldSystemdVersion()
```

Comment:
This function returns an error but we're just ignoring it here. Should it either not return an error or should we catch it here?

Reply:
Fixed.